src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -15,22 +15,21 @@
  */
 package nl.knaw.dans.easy.multideposit
 
+import nl.knaw.dans.easy.multideposit.Action.CombinedAction
+
 import scala.util.{ Failure, Success, Try }
 import nl.knaw.dans.lib.error.{ CompositeException, TraversableTryExtensions }
 import nl.knaw.dans.lib.logging.DebugEnhancedLogging
 
-import scala.collection.mutable
+import scala.annotation.tailrec
 import scala.util.control.NonFatal
 
-/**
- * An action to be performed by Process SIP. It provides three methods that can be invoked to verify
- * the feasibility of the action, to perform the action and - if necessary - to roll back the action.
- */
 /*
-  To future developers: this class is a semigroup. It satisfies the associativity law as defined
-  in https://wiki.haskell.org/Typeclassopedia#Laws_4.
+  To future developers: this class is an Applicative Functor. It satisfies the law as defined in
+    - https://wiki.haskell.org/Typeclassopedia#Laws
+    - https://wiki.haskell.org/Typeclassopedia#Laws_2.
  */
-trait Action extends DebugEnhancedLogging {
+trait Action[+T] extends DebugEnhancedLogging {
```

Comment:
This is the major change in this PR. We parameterize `Action` such that it's `run` and `execute` methods return a `Try[T]` rather than `Try[Unit]`. Note that for most subclasses of `Action` we keep `T =:= Unit`, as is shown in file diffs below.

Because of this extra parameter, we have to make a couple of changes in the report functions to make sure the compiler remains happy. Overall the report's output remained the same as before this PR.

The former `combine :: Action -> Action -> Action` combinator has been replaced with a `combineWith :: Action[T] -> Action[S] -> ((T, S) -> R) -> Action[R]` which combines two actions by providing a combinator function `(T, S) -> R`, which combines the outputs of both actions as the final result of the resulting `Action[R]`.

Furthermore we introduced convenience operators `map`, `applyLeft`, `applyRight`, `andThen` and `thenAnd`. See the scaladoc of these operators for more explanation. Notice that `x.andThen(y)` (with `x` and `y` of type `Action[Unit]`) is the same as the former `x.combine(y)`.